src/devices/src/virtio/net/device.rs

```diff
@@ -693,39 +697,268 @@ impl VirtioDevice for Net {
+
+            let guest_mac = Net::default_guest_mac();
+
+            let mut config_space;
+            config_space = Vec::with_capacity(MAC_ADDR_LEN);
+            // This is safe, because we know the capacity is large enough.
+            unsafe { config_space.set_len(MAC_ADDR_LEN) }
+            config_space[..].copy_from_slice(guest_mac.get_bytes());
+            // When this feature isn't available, the driver generates a random MAC address.
+            // Otherwise, it should attempt to read the device MAC address from the config space.
+            avail_features |= 1 << VIRTIO_NET_F_MAC;
+
+
+            let mut queue_evts = Vec::new();
+            for _ in QUEUE_SIZES.iter() {
+                queue_evts.push(EventFd::new(libc::EFD_NONBLOCK).map_err(Error::EventFd)?);
             }
+
+            let queues = QUEUE_SIZES.iter().map(|&s| Queue::new(s)).collect();
+
+            let mmds_ns = if allow_mmds_requests {
+                Some(MmdsNetworkStack::new_with_defaults())
+            } else {
+                None
+            };
+
+            Ok(Net {
+                tap,
+                avail_features,
+                acked_features: 0u64,
+                mem,
+                queues,
+                queue_evts,
+                rx_rate_limiter,
+                tx_rate_limiter,
+                rx_deferred_frame: false,
+                rx_deferred_irqs: false,
+                rx_bytes_read: 0,
+                rx_frame_buf: [0u8; MAX_BUFFER_SIZE],
+                tx_frame_buf: [0u8; MAX_BUFFER_SIZE],
+                tx_iovec: Vec::with_capacity(QUEUE_SIZE as usize),
+                interrupt_status: Arc::new(AtomicUsize::new(0)),
+                interrupt_evt: EventFd::new(libc::EFD_NONBLOCK).map_err(Error::EventFd)?,
+                guest_mac: Some(MacAddr::from_bytes_unchecked(
+                    &config_space.clone()[..MAC_ADDR_LEN],
+                )),
+                device_activated: false,
+                config_space,
+                mmds_ns,
+            })
+
+        }
+
+        fn default_guest_mac() -> MacAddr {
+            unimplemented!()
         }
     }
-}
 
-#[cfg(test)]
-mod tests {}
+    #[test]
+    fn test_vnet_helpers() {
+        let mut frame_buf: [u8; MAX_BUFFER_SIZE] = [42u8; MAX_BUFFER_SIZE];
+
+        let vnet_hdr_len_ = mem::size_of::<virtio_net_hdr_v1>();
+        assert_eq!(vnet_hdr_len_, vnet_hdr_len());
+
+        init_vnet_hdr(&mut frame_buf);
+        let zero_vnet_hdr = vec![0u8; vnet_hdr_len_];
+        assert_eq!(zero_vnet_hdr, &frame_buf[..vnet_hdr_len_]);
+
+        let payload = vec![42u8; MAX_BUFFER_SIZE - vnet_hdr_len_];
+        assert_eq!(payload, frame_bytes_from_buf(&frame_buf));
+
+        {
+            let payload = frame_bytes_from_buf_mut(&mut frame_buf);
+            payload[0] = 15;
+        }
+        assert_eq!(frame_buf[vnet_hdr_len_], 15);
+    }
+
+    #[test]
+    #[allow(clippy::cognitive_complexity)]
```

Comment:
Instead of allowing cognitive complexity, we should create smaller tests. From a quick look it seems like each block of code can be in its separate test as they don't depend on previous configurations.